### ISSUE TITLE
Vite template doesn't detect the `.server` modules

### ISSUE THREAD

**Issue by wahyubucil** [remix-run/remix#8898]
### Reproduction

Go to https://stackblitz.com/edit/remix-run-remix-g5iaoj (this is a clone of the Vite template).
Run `pnpm run typecheck`

We will have 2 issues:

First, we will have an error: `error TS2688: Cannot find type definition file for 'node'.`. 
Do we need the `node` types declaration? If yes, then we need to include `@types/node` in the `devDependencies` of package.json. If not, then we need to remove `node` from the `types` on `tsconfig.json`.

Second, if we've already fixed the issue above, then look at the file `app/.server/example.ts`. It has an error, but when we're running the `pnpm run typecheck`, it doesn't detect the error. The reason is that TypeScript by default is [ignoring hidden files](https://github.com/microsoft/TypeScript/issues/49555).

### System Info

```shell
This is Stackblitz system info:

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.14.0 - /usr/local/bin/pnpm
  npmPackages:
    @remix-run/dev: ^2.7.2 => 2.7.2 
    @remix-run/node: ^2.7.2 => 2.7.2 
    @remix-run/react: ^2.7.2 => 2.7.2 
    @remix-run/serve: ^2.7.2 => 2.7.2 
    vite: ^5.1.4 => 5.1.4
```

### Used Package Manager

pnpm

### Expected Behavior

TypeScript should detect errors on the `.server` directories.

### Actual Behavior

It's ignoring the `.server` directories.

**Comment by kiliman** [comment]
For _.server_ folders, you'll need to explicitly add it to the `include` array in _tsconfig.json_. This only applies to files/folders starting with `.` It will work if you name it something like `models.server`.

```json
{
  "include": ["env.d.ts", "**/*.ts", "**/*.tsx", "app/.server/**/*.ts"],
  "compilerOptions": {
  }
}
```

https://github.com/microsoft/TypeScript/issues/49555

As for Node types, that's interesting. In my `remix-vite-template`, I don't include the `types` option in my _tsconfig.json_ file. I asked ChatGPT what it thought, and the Cons outweigh the Pros, in my opinion.

## 🤖 Ask ChatGPT 4

### ❓ What are the pros and cons of including the `types` option in _tsconfig.json_?

Including the `types` option in _tsconfig.json_ for a TypeScript project can significantly impact how the TypeScript compiler recognizes and uses type definitions in your project. Here are the pros and cons of using this option:

### Pros:

1. **Explicit Control Over Available Global Types**: By specifying which types are included, you can limit the global type declarations that are available in your project. This is particularly useful in large projects or when using multiple libraries that might define conflicting global types.

2. **Performance Improvement**: Limiting the types loaded into your project can lead to faster compilation times, as the TypeScript compiler has fewer type definitions to process. This is especially noticeable in projects with many type declaration files.

3. **Avoiding Type Conflicts**: In projects where multiple libraries might define similar global types, specifying which types to include can help avoid conflicts and ambiguous type errors. This makes your type environment more predictable and stable.

4. **Better Code Autocompletion**: By controlling which types are available, you can ensure that the code autocompletion features in your IDE are more relevant and less cluttered with types from libraries that are not used in parts of your project.

### Cons:

1. **Manual Management**: You need to manually manage which types are included in your project. This can become cumbersome, especially as your project grows or as you add and remove dependencies.

2. **Risk of Missing Types**: If you forget to include a necessary type definition, you might encounter hard-to-debug errors due to the missing type information. This requires a good understanding of which types your project actually needs.

3. **Overhead in Large Projects**: For large projects with many dependencies, configuring and maintaining the `types` option can become a significant overhead. This might require regular updates to _tsconfig.json_ as dependencies evolve.

4. **Potential for Incomplete Type Safety**: Excluding certain types can lead to a false sense of security if parts of your codebase rely on types that are not included. This might result in parts of your code not being fully type-checked.

In summary, using the `types` option in _tsconfig.json_ gives you more control over your project's type environment, potentially leading to performance improvements and better developer experience. However, it requires careful management and a good understanding of your project's dependencies to avoid missing or conflicting types. The decision to use this option should be based on the specific needs and complexity of your project.

https://chat.openai.com/share/892bd94c-076c-4a5d-9224-f7feeca3d794
